File: bindings/java/src/main/java/io/slatedb/WriteHandle.java
package io.slatedb;


import java.util.List;
import java.util.Map;
import java.nio.ByteBuffer;
import java.util.Objects;
/**
 * Metadata returned from a successful write.
 */
public class WriteHandle {
    /**
     * The sequence number assigned to the write.
     */
    private Long seqnum;
    /**
     * The creation timestamp assigned to the write.
     */
    private Long createTs;

    public WriteHandle(
        Long seqnum, 
        Long createTs
    ) {
        
        this.seqnum = seqnum;
        
        this.createTs = createTs;
    }
    
    public Long seqnum() {
        return this.seqnum;
    }
    
    public Long createTs() {
        return this.createTs;
    }
    public void setSeqnum(Long seqnum) {
        this.seqnum = seqnum;
    }
    public void setCreateTs(Long createTs) {
        this.createTs = createTs;
    }

    
    
    @Override
    public boolean equals(Object other) {
        if (other instanceof WriteHandle) {
            WriteHandle t = (WriteHandle) other;
            return (
              Objects.equals(seqnum, t.seqnum) && 
              
              Objects.equals(createTs, t.createTs)
              
            );
        };
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(seqnum, createTs);
    }
}

File: bindings/uniffi/src/write_handle.rs
use crate::error::Error;

/// Handle returned by a successful write.
#[derive(uniffi::Object)]
pub struct WriteHandle {
    inner: slatedb::WriteHandle,
}

impl WriteHandle {
    pub(crate) fn new(inner: slatedb::WriteHandle) -> Self {
        Self { inner }
    }
}

#[uniffi::export]
impl WriteHandle {
    /// Returns the sequence number assigned to the write.
    pub fn seqnum(&self) -> u64 {
        self.inner.seqnum()
    }

    /// Returns the creation timestamp assigned to the write.
    pub fn create_ts(&self) -> i64 {
        self.inner.create_ts()
    }
}

#[uniffi::export(async_runtime = "tokio")]
impl WriteHandle {
    /// Waits until the write has been durably persisted.
    pub async fn await_durable(&self) -> Result<(), Error> {
        self.inner.await_durable().await.map_err(Into::into)
    }
}
